splunkhecexporter panic

### Component(s)

exporter/splunkhec

### What happened?

## Description

Our otel-collector panicked with the following stacktrace:

```
	/usr/local/go/src/net/http/transport.go:1800 +0x1585
created by net/http.(*Transport).dialConn in goroutine 176037
	/usr/local/go/src/net/http/transport.go:2447 +0x18e
net/http.(*persistConn).writeLoop(0xc0012d9e60)
	/usr/local/go/src/net/http/request.go:755 +0xb0d
net/http.(*Request).write(0xc0015c3c20, {0x1962520, 0xc0003ded00}, 0x0, 0xc001877ce0, 0x0)
	/usr/local/go/src/net/http/transfer.go:385 +0x426
net/http.(*transferWriter).writeBody(0xc0005c60a0, {0x1962520?, 0xc0003ded00?})
	/usr/local/go/src/net/http/transport.go:662 +0x22
net/http.(*readTrackingBody).Close(0xc0005c60a0?)
	/go/pkg/mod/go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp@v0.52.0/wrap.go:37 +0x1b
go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp.(*bodyWrapper).Close(0xc0017b4be0?)
	/go/pkg/mod/github.com/open-telemetry/opentelemetry-collector-contrib/exporter/splunkhecexporter@v0.104.0/buffer.go:117 +0x17
github.com/open-telemetry/opentelemetry-collector-contrib/exporter/splunkhecexporter.(*cancellableGzipWriter).Close(0xc000918000?)
	/usr/local/go/src/compress/gzip/gzip.go:242 +0x7a
compress/gzip.(*Writer).Close(0xc001298000)
	/usr/local/go/src/compress/flate/deflate.go:730
compress/flate.(*Writer).Close(...)
	/usr/local/go/src/compress/flate/deflate.go:638 +0x6a
compress/flate.(*compressor).close(0xc001f2c000)
	/usr/local/go/src/compress/flate/deflate.go:408 +0x618
compress/flate.(*compressor).deflate(0xc001f2c000)
	/usr/local/go/src/compress/flate/deflate.go:170 +0x9d
compress/flate.(*compressor).writeBlock(0xc001f2c000, {0xc001552000?, 0xc001918f01?, 0xc000473bc0?}, 0xc0017b4b50?)
	/usr/local/go/src/compress/flate/huffman_bit_writer.go:437 +0xbe
compress/flate.(*huffmanBitWriter).writeBlock(0xc0005e32c0, {0xc001552000?, 0x80?, 0x80?}, 0x0, {0xc00140b1d7, 0x0, 0xee29})
	/usr/local/go/src/compress/flate/huffman_bit_writer.go:548 +0x2a5
compress/flate.(*huffmanBitWriter).indexTokens(0xc0005e32c0, {0xc001552000, 0x4e0, 0x47ecbd?})
goroutine 175868 [running]:
panic: runtime error: index out of range [43] with length 30
```

I suspect a race condition (it rarely occurs), and possibly a variation of the issue previously reported here https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/34255 But here the symptom is different/more sever. 

### Collector version

v0.104.0

### Environment information

## Environment
alpine linux
Go 1.22


### OpenTelemetry Collector configuration

```yaml
exporters:
  splunk_hec/1:
    token: "{{.hecToken}}"
    endpoint: "{{.hecEndpoint}}/services/collector/raw?index={{.splunkIndex}}&sourcetype={{.splunkSourceType}}&source=eventhub://{{.eventHubFullyQualifiedNamespace}}/{{.eventHubName}}
    sourcetype: "{{.splunkSourceType}}"
    index: "{{.splunkIndex}}"
    export_raw: true
    max_content_length_logs: {{.exporterMaxContentLengthLogs}}
    retry_on_failure:
      enabled: false
    sending_queue:
      enabled: false
    hec_metadata_to_otel_attrs:
      source: source
      host: host
    tls:
      min_version: 1.2
```


### Log output

_No response_

### Additional context

_No response_

--- a/.chloggen/safe_bytes.yaml
+++ b/.chloggen/safe_bytes.yaml
@@ -1,0 +1,31 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: bug_fix
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: splunkhecexporter
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Copy the bytes to be placed in the request body to avoid corruption on reuse
+
+# Mandatory: One or more tracking issues related to the change. You can use the PR number here if no issue exists.
+issues: [34357]
+
+# (Optional) One or more lines of additional information to render under the primary note.
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext: |
+  This bug is a manifestation of https://github.com/golang/go/issues/51907.
+  Under high load, the pool of buffers used to send requests is reused enough
+  that the same buffer is used concurrently to process data and be sent as request body.
+  The fix is to copy the payload into a new byte array before sending it.
+
+# If your change doesn't affect end users or the exported elements of any package,
+# you should instead start your pull request title with [chore] or use the "Skip Changelog" label.
+# Optional: The change log or logs in which this entry should be included.
+# e.g. '[user]' or '[user, api]'
+# Include 'user' if the change is relevant to end users.
+# Include 'api' if there is a change to a library API.
+# Default: '[user]'
+change_logs: []

--- a/exporter/splunkhecexporter/buffer.go
+++ b/exporter/splunkhecexporter/buffer.go
@@ -22,6 +22,7 @@ type buffer interface {
 	Reset()
 	Len() int
 	Empty() bool
+	Bytes() []byte
 }
 
 type cancellableBytesWriter struct {
@@ -57,6 +58,10 @@ func (c *cancellableBytesWriter) Len() int {
 
 func (c *cancellableBytesWriter) Empty() bool {
 	return c.innerWriter.Len() == 0
+}
+
+func (c *cancellableBytesWriter) Bytes() []byte {
+	return c.innerWriter.Bytes()
 }
 
 type cancellableGzipWriter struct {
@@ -123,6 +128,10 @@ func (c *cancellableGzipWriter) Len() int {
 
 func (c *cancellableGzipWriter) Empty() bool {
 	return c.rawLen == 0
+}
+
+func (c *cancellableGzipWriter) Bytes() []byte {
+	return c.innerBuffer.Bytes()
 }
 
 // bufferPool is a pool of buffer objects.

--- a/exporter/splunkhecexporter/hec_worker.go
+++ b/exporter/splunkhecexporter/hec_worker.go
@@ -28,6 +28,7 @@ type defaultHecWorker struct {
 }
 
 func (hec *defaultHecWorker) send(ctx context.Context, buf buffer, headers map[string]string) error {
+	// We copy the bytes to a new buffer to avoid corruption. This is a workaround to avoid hitting https://github.com/golang/go/issues/51907.
 	nb := make([]byte, buf.Len())
 	copy(nb, buf.Bytes())
 	bodyBuf := bytes.NewReader(nb)

--- a/exporter/splunkhecexporter/hec_worker.go
+++ b/exporter/splunkhecexporter/hec_worker.go
@@ -4,6 +4,7 @@
 package splunkhecexporter // import "github.com/open-telemetry/opentelemetry-collector-contrib/exporter/splunkhecexporter"
 
 import (
+	"bytes"
 	"context"
 	"io"
 	"net/http"
@@ -27,7 +28,10 @@ type defaultHecWorker struct {
 }
 
 func (hec *defaultHecWorker) send(ctx context.Context, buf buffer, headers map[string]string) error {
-	req, err := http.NewRequestWithContext(ctx, "POST", hec.url.String(), buf)
+	nb := make([]byte, buf.Len())
+	copy(nb, buf.Bytes())
+	bodyBuf := bytes.NewReader(nb)
+	req, err := http.NewRequestWithContext(ctx, "POST", hec.url.String(), bodyBuf)
 	if err != nil {
 		return consumererror.NewPermanent(err)
 	}

